pkg/volume/csi/csi_attacher.go

```diff
@@ -235,18 +271,34 @@ func (c *csiAttacher) VolumesAreAttached(specs []*volume.Spec, nodeName types.No
 	attached := make(map[*volume.Spec]bool)
 
 	for _, spec := range specs {
+		var driver, handle string
 		if spec == nil {
 			klog.Error(log("attacher.VolumesAreAttached missing volume.Spec"))
 			return nil, errors.New("missing spec")
 		}
-		source, err := getCSISourceFromSpec(spec)
+
+		volSrc, pvSrc, err := getSourceFromSpec(spec)
 		if err != nil {
 			klog.Error(log("attacher.VolumesAreAttached failed: %v", err))
 			continue
 		}
-		skip, err := c.plugin.skipAttach(source.Driver)
+
+		if c.plugin.inlineFeatureEnabled && volSrc != nil {
```

Comment:
seems like this block (or something very similar to it) is repeated in several places: checking whether its vol or pv source then getting the driver and handle out. Can this be factored into a common place?